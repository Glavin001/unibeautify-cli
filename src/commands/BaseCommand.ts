import chalk from "chalk";

export abstract class BaseCommand {
  protected stdout: NodeJS.WriteStream = process.stdout;
  protected stderr: NodeJS.WriteStream = process.stderr;

  protected writeOut(text: string): void {
    this.stdout.write(text + "\n");
  }

  protected writeOutHeading(text: string): void {
    this.writeOut(this.stdout.isTTY ? chalk.blue(text) : text);
  }

  protected writeError(text: string): void {
    this.stderr.write((this.stderr.isTTY ? chalk.red(text) : text) + "\n");
  }

  protected set exitCode(exitCode: number) {
    process.exitCode = exitCode;
  }

  protected get exitCode(): number {
    return process.exitCode || 0;
  }

  protected handleError(error: Error, exitCode: number): Promise<never> {
    this.writeError(error.message);
    this.exitCode = exitCode;
    return Promise.reject(error);
  }
}
