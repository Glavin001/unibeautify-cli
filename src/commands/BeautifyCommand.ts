import { BeautifyData } from "unibeautify";
import * as cosmiconfig from "cosmiconfig";
import * as fs from "fs";

import { setupUnibeautify } from "../utils";
import { BaseCommand } from "./BaseCommand";

export class BeautifyCommand extends BaseCommand {
  protected stdin: NodeJS.ReadStream = process.stdin;

  public beautify(programArgs: IArgs): Promise<string> {
    const { language, filePath } = programArgs;
    return setupUnibeautify().then(unibeautify => {
      if (!language) {
        const error = new Error("A language is required.");
        this.writeError(error.message);
        this.exitCode = 1;
        return Promise.reject(error);
      }
      return Promise.all([
        this.readConfig(programArgs),
        this.readText(filePath),
      ]).then(([config, text]) => {
        const data: BeautifyData = {
          filePath: filePath,
          languageName: language,
          options: (config as any) || {},
          text,
        };
        return unibeautify
          .beautify(data)
          .then((result: string) => {
            this.writeOut(result);
            return result;
          })
          .catch((error: Error) => {
            this.writeError(error.message);
            this.exitCode = 1;
            return Promise.reject(error);
          });
      });
    });
  }

  private readConfig(programArgs: IArgs): Promise<any> {
    const { configFile, configJson } = programArgs;
    if (configJson) {
      return this.parseConfig(configJson);
    } else {
      return this.configFile(configFile);
    }
  }

  private readText(filePath?: string): Promise<string> {
    // if (this.isTerminal && filePath) {
    if (filePath) {
      return this.readFile(filePath);
    } else {
      return this.readFromStdin();
    }
  }

  private parseConfig(configJson: string): Promise<object> {
    try {
      return Promise.resolve(JSON.parse(configJson));
    } catch (error) {
      this.writeError(error.message);
      this.exitCode = 2;
      return Promise.reject(error);
    }
  }

  private configFile(configFile?: string, filePath?: string) {
    const configExplorer = cosmiconfig("unibeautify", {});
    const loadConfigPromise = configFile
      ? configExplorer.load(configFile)
      : configExplorer.search(filePath);
    return loadConfigPromise
      .then(result => (result ? result.config : null))
      .catch(error =>
        Promise.reject(
          new Error(`Could not load configuration file ${configFile}`)
        )
      );
  }

  protected get isTerminal(): boolean {
    return Boolean(this.stdin.isTTY);
  }

  protected readFromStdin(): Promise<string> {
    return new Promise((resolve, reject) => {
      let text = "";
      this.stdin.on("data", (data: string) => {
        text = data.toString();
      });
      this.stdin.on("end", () => {
        return resolve(text);
      });
      this.stdin.on("error", (err: any) => {
        if (err.code === "EPIPE") {
          this.exitCode = 1;
          return reject(err);
        }
        process.emit("warning", err);
      });
      this.stdin.resume();
    });
  }

  private readFile(filePath: string): Promise<string> {
    return new Promise((resolve, reject) => {
      fs.readFile(filePath, (error, data) => {
        if (error) {
          return reject(error);
        }
        return resolve(data.toString());
      });
    });
  }
}

/**
  Arguments parsed for program.
  */
export interface IArgs {
  args: string[];
  language?: string;
  outFile?: string;
  replace?: boolean;
  configFile?: string;
  configJson?: string;
  filePath?: string;
}
