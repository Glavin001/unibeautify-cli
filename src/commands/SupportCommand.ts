import {
  getSupportedLanguages,
  findInstalledBeautifiers,
  setupUnibeautify,
  getAllLanguages,
} from "../utils";
import { BaseCommand } from "./BaseCommand";

export class SupportCommand extends BaseCommand {
  public support(options: {
    json?: boolean;
    languages?: boolean;
    beautifiers?: boolean;
    all?: boolean;
  }): Promise<void> {
    const printer: (info: SupportInfo) => void = options.json
      ? this.jsonPrinter
      : this.listPrinter;
    const info: SupportInfo = {};
    return setupUnibeautify().then(async unibeautify => {
      if (options.languages) {
        if (options.all) {
          info["languages"] = getAllLanguages();
        } else {
          info["languages"] = getSupportedLanguages();
        }
      }
      if (options.beautifiers) {
        const beautifiers = await findInstalledBeautifiers();
        info["beautifiers"] = beautifiers;
      }
      if (Object.keys(info).length === 0) {
        this.writeError("Nothing to show");
        // this.exit(1);
        this.exitCode = 1;
      } else {
        printer(info);
        // this.exit(0);
      }
    });
  }

  private jsonPrinter = (info: SupportInfo) => {
    this.writeOut(JSON.stringify(info, null, 2));
    // unibeautify:ignore-next-line
  }

  private listPrinter = (info: SupportInfo) => {
    Object.keys(info).forEach(section => {
      this.writeOutHeading(`Supported ${section}`);
      const items = info[section];
      items.forEach((item, index) => this.writeOut(`${index + 1}. ${item}`));
    });
    // unibeautify:ignore-next-line
  }
}

interface SupportInfo {
  [section: string]: string[];
}
