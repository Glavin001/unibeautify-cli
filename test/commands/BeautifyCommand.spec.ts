import { BeautifyCommand } from "../../src/commands/BeautifyCommand";
import {
  createMockWritableStream,
  createMockReadableStream,
} from "../mockStreams";
import * as fs from "fs";

describe("BeautifyCommand", () => {
  class CustomCommand extends BeautifyCommand {
    constructor(private rawStdin: string = "", private isTTY: boolean = false) {
      super();
      this.rawStdin;
    }
    protected stdin = createMockReadableStream(this.rawStdin, this.isTTY);
    protected stdout = createMockWritableStream();
    protected stderr = createMockWritableStream();
    public toJSON() {
      return {
        exitCode: this.exitCode,
        stderr: this.stderr.rawData,
        stdout: this.stdout.rawData,
      };
    }
  }
  describe("Beautify", () => {
    test("should beautify using options passed in cmd", async () => {
      const command = new CustomCommand();
      await command.beautify({
        args: [],
        configJson: `{"JavaScript": {"beautifiers": ["ESLint"],"quotes": "double"}}`,
        filePath: "test/fixtures/test1.js",
        language: "JavaScript",
      });
      const json = command.toJSON();
      expect(json.exitCode).toBe(0);
      expect(json).toMatchSnapshot();
    });
    test("should beautify using options in config file", async () => {
      const command = new CustomCommand();
      await command.beautify({
        args: [],
        configFile: "test/.unibeautifyrc.yml",
        filePath: "test/fixtures/test1.js",
        language: "JavaScript",
      });
      const json = command.toJSON();
      expect(json.exitCode).toBe(0);
      expect(json).toMatchSnapshot();
    });
    test("should find config file without specifying it", async () => {
      const command = new CustomCommand();
      await command.beautify({
        args: [],
        filePath: "test/fixtures/test1.js",
        language: "JavaScript",
      });
      const json = command.toJSON();
      expect(json.exitCode).toBe(0);
      expect(json).toMatchSnapshot();
    });
    test("should beautify using stdin", async () => {
      const command = new CustomCommand("const test = 'test';");
      return command
        .beautify({
          args: [],
          configFile: "test/.unibeautifyrc.yml",
          language: "JavaScript",
        })
        .then(() => {
          const json = command.toJSON();
          expect(json.exitCode).toBe(0);
          expect(json.stderr).toBe("");
          // tslint:disable-next-line:quotemark
          expect(json.stdout).toBe('const test = "test";\n');
        });
    });
    test("should beautify using stdin with TTY", async () => {
      const command = new CustomCommand("const test = 'test';", true);
      return command
        .beautify({
          args: [],
          configFile: "test/.unibeautifyrc.yml",
          language: "JavaScript",
        })
        .then(() => {
          const json = command.toJSON();
          expect(json.exitCode).toBe(0);
          expect(json.stderr).toBe("");
          // tslint:disable-next-line:quotemark
          // expect(json.stdout).toBe('const test = "test";\n'); // <--- Doesn't work
          expect(json.stdout).toBe("\n"); // <--- What it actually returns
        });
    });
    test("should accept blank files or text", async () => {
      const command = new CustomCommand();
      return command
        .beautify({
          args: [],
          configFile: "test/.unibeautifyrc.yml",
          language: "JavaScript",
        })
        .then(() => {
          const json = command.toJSON();
          expect(json.exitCode).toBe(0);
          expect(json.stderr).toBe("");
          // tslint:disable-next-line:quotemark
          expect(json.stdout).toBe("\n");
        });
    });
    test("should beautify and write to file", async () => {
      const command = new CustomCommand("const test = 'test';");
      const originPath = "test/fixtures/test1.js";
      const destPath = "test/commands/test1.js";
      return copyFile(originPath, destPath).then(() => {
        return command
          .beautify({
            args: [],
            configFile: "test/.unibeautifyrc.yml",
            filePath: destPath,
            inplace: true,
            language: "JavaScript",
          })
          .then(() => {
            return readFile(destPath).then(result => {
              const json = command.toJSON();
              expect(json.exitCode).toBe(0);
              expect(json.stderr).toBe("");
              // tslint:disable-next-line:quotemark
              expect(result.toString()).toBe('const test = "test";\n');
              return unlink(destPath);
            });
          });
      });
    });
  });
  describe("Errors", () => {
    test("should throw error when cannot find text file", () => {
      const command = new CustomCommand();
      const thenCb = jest.fn();
      const catchCb = jest.fn();
      return command
        .beautify({
          args: [],
          configFile: "test/.unibeautifyrc.yml",
          filePath: "test/test2.js",
          language: "JavaScript",
        })
        .then(thenCb)
        .catch(catchCb)
        .then(() => {
          expect(thenCb).not.toBeCalled();
          expect(catchCb).toHaveBeenCalled();
          expect(catchCb.mock.calls).toHaveLength(1);
          expect(catchCb.mock.calls[0]).toHaveLength(1);
          expect((<any>catchCb.mock.calls[0][0]).message).toBe(
            "ENOENT: no such file or directory, open 'test/test2.js'"
          );
        });
    });
    test("should throw error when cannot find config", () => {
      const command = new CustomCommand();
      const thenCb = jest.fn();
      const catchCb = jest.fn();
      const configFile = "test/.unibeautifyrc2.yml";
      return command
        .beautify({
          args: [],
          configFile,
          filePath: "test/fixtures/test1.js",
          language: "JavaScript",
        })
        .then(thenCb)
        .catch(catchCb)
        .then(() => {
          expect(thenCb).not.toBeCalled();
          expect(catchCb).toHaveBeenCalled();
          expect(catchCb.mock.calls).toHaveLength(1);
          expect(catchCb.mock.calls[0]).toHaveLength(1);
          expect((<any>catchCb.mock.calls[0][0]).message).toBe(
            `Could not load configuration file ${configFile}`
          );
        });
    });
    test("should throw an error saying language is required", () => {
      const command = new CustomCommand();
      const thenCb = jest.fn();
      const catchCb = jest.fn();
      return command
        .beautify({
          args: [],
          configFile: "test/.unibeautifyrc.yml",
          filePath: "test/fixtures/test1.js",
        })
        .then(thenCb)
        .catch(catchCb)
        .then(() => {
          expect(thenCb).not.toBeCalled();
          expect(catchCb).toHaveBeenCalled();
          expect(catchCb.mock.calls).toHaveLength(1);
          expect(catchCb.mock.calls[0]).toHaveLength(1);
          expect(catchCb).toHaveProperty(
            ["mock", "calls", 0, 0, "message"],
            "A language is required."
          );
          const json = command.toJSON();
          expect(json.exitCode).toBe(1);
          expect(json).toMatchSnapshot("json");
        });
    });
    test("should throw an error saying cannot find language", () => {
      const command = new CustomCommand();
      const thenCb = jest.fn();
      const catchCb = jest.fn();
      return command
        .beautify({
          args: [],
          configFile: "test/.unibeautifyrc.yml",
          filePath: "test/fixtures/test1.js",
          language: "javascript",
        })
        .then(thenCb)
        .catch(catchCb)
        .then(() => {
          expect(thenCb).not.toBeCalled();
          expect(catchCb).toHaveBeenCalled();
          expect(catchCb.mock.calls).toHaveLength(1);
          expect(catchCb.mock.calls[0]).toHaveLength(1);
          expect(catchCb).toHaveProperty(
            ["mock", "calls", 0, 0, "message"],
            "Cannot find language."
          );
          const json = command.toJSON();
          expect(json.exitCode).toBe(1);
          expect(json).toMatchSnapshot("json");
        });
    });
    test("should throw an error with invalid json", () => {
      const command = new CustomCommand();
      const thenCb = jest.fn();
      const catchCb = jest.fn();
      return command
        .beautify({
          args: [],
          configJson: `{"JavaScript": {"beautifiers": ["ESLint"],"quotes": "double"`,
          filePath: "test/fixtures/test1.js",
          language: "JavaScript",
        })
        .then(thenCb)
        .catch(catchCb)
        .then(() => {
          expect(thenCb).not.toBeCalled();
          expect(catchCb).toHaveBeenCalled();
          expect(catchCb.mock.calls).toHaveLength(1);
          expect(catchCb.mock.calls[0]).toHaveLength(1);
          expect(catchCb).toHaveProperty(
            ["mock", "calls", 0, 0, "message"],
            "Unexpected end of JSON input"
          );
          const json = command.toJSON();
          expect(json.exitCode).toBe(2);
          expect(json).toMatchSnapshot("json");
        });
    });
  });
});

function copyFile(filePath: string, destPath: string): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.copyFile(filePath, destPath, error => {
      if (error) {
        return reject(error);
      }
      return resolve();
    });
  });
}

function readFile(filePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    fs.readFile(filePath, (error, data) => {
      if (error) {
        return reject(error);
      }
      return resolve(data.toString());
    });
  });
}

function unlink(filePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    fs.unlink(filePath, error => {
      if (error) {
        return reject(error);
      }
      return resolve();
    });
  });
}
