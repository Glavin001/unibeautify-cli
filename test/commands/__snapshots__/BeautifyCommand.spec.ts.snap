// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeautifyCommand Beautify should beautify using options in config file 1`] = `
Object {
  "exitCode": 0,
  "stderr": "",
  "stdout": "const test = \\"test\\";

",
}
`;

exports[`BeautifyCommand Beautify should beautify using options passed in cmd 1`] = `
Object {
  "exitCode": 0,
  "stderr": "",
  "stdout": "const test = \\"test\\";

",
}
`;

exports[`BeautifyCommand Beautify should find config file without specifying it 1`] = `
Object {
  "exitCode": 0,
  "stderr": "",
  "stdout": "const test = \\"test\\";

",
}
`;

exports[`BeautifyCommand Errors should throw an error saying language is required: json 1`] = `
Object {
  "exitCode": 1,
  "stderr": "A language is required.
",
  "stdout": "",
}
`;

exports[`BeautifyCommand Errors should throw an error with invalid json: json 1`] = `
Object {
  "exitCode": 2,
  "stderr": "Unexpected end of JSON input
",
  "stdout": "",
}
`;
